executor/stale_txn_test.go

```diff
@@ -116,6 +117,104 @@ func (s *testStaleTxnSerialSuite) TestExactStalenessTransaction(c *C) {
 	}
 }
 
+func (s *testStaleTxnSerialSuite) TestSelectAsOf(c *C) {
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/executor/mockStalenessTxnSchemaVer", "return(false)"), IsNil)
+	defer func() {
+		err := failpoint.Disable("github.com/pingcap/tidb/executor/mockStalenessTxnSchemaVer")
+		c.Assert(err, IsNil)
+	}()
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec(`drop table if exists b`)
+	tk.MustExec("create table t (id int primary key);")
+	tk.MustExec("create table b (id int primary key);")
+	defer func() {
+		tk.MustExec(`drop table if exists b`)
+		tk.MustExec(`drop table if exists t`)
+	}()
+
+	testcases := []struct {
+		name             string
+		sql              string
+		expectPhysicalTS int64
+		preSec           int64
+		// IsStaleness is auto cleanup in select stmt.
+		errorStr string
+	}{
+		{
+			name:             "TimestampExactRead",
+			sql:              `select * from t as of timestamp '2020-09-06 00:00:00';`,
+			expectPhysicalTS: 1599321600000,
+		},
+		{
+			name:             "TimestampExactRead",
+			sql:              `select * from t as of timestamp TIMESTAMP('2020-09-06 00:00:00');`,
+			expectPhysicalTS: 1599321600000,
+		},
+		{
+			name:   "TimestampExactRead",
+			sql:    `select * from t as of timestamp NOW() - INTERVAL 20 SECOND;`,
+			preSec: 20,
+		},
+		{
+			name:   "TimestampExactRead",
+			sql:    `select * from t as of timestamp TIMESTAMP(NOW() - INTERVAL 20 SECOND);`,
+			preSec: 20,
+		},
+		{
+			name:   "TimestampExactRead",
+			sql:    `select * from t as of timestamp TIMESTAMP(NOW() - INTERVAL 20 SECOND), b as of timestamp TIMESTAMP(NOW() - INTERVAL 20 SECOND);`,
+			preSec: 20,
+		},
+		{
+			name:     "TimestampExactRead",
+			sql:      `select * from t as of timestamp TIMESTAMP(NOW() - INTERVAL 20 SECOND), b as of timestamp TIMESTAMP('2020-09-06 00:00:00');`,
+			preSec:   20,
+			errorStr: "not set different timestamp",
+		},
+		{
+			name:     "TimestampExactRead",
+			sql:      `select * from t as of timestamp TIMESTAMP(NOW() - INTERVAL 20 SECOND), b;`,
+			preSec:   20,
+			errorStr: "not set different timestamp",
+		},
+		{
+			name:     "TimestampExactRead",
+			sql:      `select * from t, b as of timestamp TIMESTAMP(NOW() - INTERVAL 20 SECOND);`,
+			preSec:   20,
+			errorStr: "not set different timestamp",
+		},
+		{
+			name:   "NomalRead",
+			sql:    `select * from t, b;`,
+			preSec: 0,
+		},
+	}
+
+	tk.MustExec("use test")
+	for _, testcase := range testcases {
+		c.Log(testcase.name)
+		_, err := tk.Exec(testcase.sql)
+		if len(testcase.errorStr) == 0 {
+			c.Assert(err, IsNil, Commentf("sql:%s, error stack %v", testcase.sql, errors.ErrorStack(err)))
+		} else {
+			c.Assert(err, NotNil)
+			c.Assert(strings.Contains(err.Error(), testcase.errorStr), IsTrue)
+			continue
+		}
```

Comment:
Maybe exchange two branches? Then `IsNil` branch will be closer to the examination below.